include/pybind11/pybind11.h

```diff
@@ -1569,6 +1569,59 @@ inline str enum_name(handle arg) {
     return "???";
 }
 
+class enum_meta_info {
+public:
+    static pybind11::object enum_meta_cls() {
+        return get().enum_meta_cls_;
+    }
+
+    static pybind11::object enum_base_cls() {
+        return get().enum_base_cls_;
+    }
+
+private:
+    template <typename T>
+    friend T& pybind11::get_or_create_shared_data(const std::string&);
+
+    static const enum_meta_info& get() {
+        return pybind11::get_or_create_shared_data<enum_meta_info>(
+            "_pybind11_enum_meta_info");
+    }
+
+    enum_meta_info() {
+        handle copy = pybind11::module::import("copy").attr("copy");
+        locals_ = copy(pybind11::globals());
+        locals_["pybind11_meta_cls"] = reinterpret_borrow<object>(
+            reinterpret_cast<PyObject*>(get_internals().default_metaclass));
+        locals_["pybind11_base_cls"] = reinterpret_borrow<object>(
+            get_internals().instance_base);
+        // TODO: Make the base class work.
+        const char code[] = R"""(
+pybind11_enum_base_cls = None
+
+class pybind11_enum_meta_cls(pybind11_meta_cls):
+    is_pybind11_enum = True
+
+    def __iter__(cls):
+        return iter(cls.__members__.values())
+
+    def __len__(cls):
+        return len(cls.__members__)
```

Comment:
> does not support isinstance(pybind11_enum_value, enum.Enum)

Is there a particular reason to not implement `__instancecheck__` here?

Reply:
Sorry, didn't realize this should be done in `enum.EnumMeta`, not here. :facepalm: 

Is there a way to do it without altering builtin classes?

Reply:
Not that I'm aware of :( I had gone the route of modifying builtin classes, but *yuck*:
https://github.com/pybind/pybind11/compare/master...a41944c57509fa3845ef190a84972811a87eac80

Reply:
I agree, it does look too hacky and doesn't sound like right thing to do. 

Actually pybind-enum is NOT `enum.Enum`, so maybe we shouldn't try to make this check return true? I think it would be more appropriate to introduce `collections.abc.Enum` (e.g. in python 3.10) and make `isinstance(pybind_enum_value, colletions.abc.Enum)` recognize all PEP-complaining enum implementations? In meantime one probably have no choice other than stick to self-written check method (like one in tests in this PR). 

I think such doubtful hacks would greatly reduce the PR merge chances, so I would leave attempts to introduce them, at least for now.